File: src/leetcode/WordSearch.java
package leetcode;

public class WordSearch {
	
	
	/**
	 * Given a 2D board and a word, find if the word exists in the grid.
	 * @param board
	 * @param word
	 * @return
	 */
	public boolean exist(char[][] board, String word) {
		
		if(board == null || board.length == 0 || board[0].length == 0)
			return false;
		
		if(word == null || word.length() == 0){
			return true;
		}

		int N_X = board.length;
		int N_Y = board[0].length;
		boolean [][] visited = new boolean[N_X][N_Y];

		for(int x = 0 ;x <  N_X; x++){
			for(int y = 0; y < N_Y; y++){
				
				if(exists(x, y, 0, board,visited, word) ){
					return true;
				}
			}
		}
		return false;
	}
	
	boolean exists(int x, int y, int loc, char[][] board, boolean [][] visited, String word){
		
		visited[x][y] = true;
		
		if(board[x][y] != word.charAt(loc)){
			return false;
		}
		
		if(loc == word.length() - 1){
 			return true;			 
		}
		
		if( x > 0 && !visited[x- 1][y] && exists(x - 1, y, loc + 1, board, visited, word)){
			return true;
		}
		
		if( y > 0 && !visited[x][y-1] && exists(x, y - 1, loc + 1, board, visited, word)){
			return true;
		}
		if( x < board.length - 1 && !visited[x + 1][y] && exists(x + 1, y, loc + 1, board, visited, word)){
			return true;
		}
		if( y < board[0].length - 1 && !visited[x][y+1] && exists(x, y + 1, loc + 1, board, visited, word)){
			return true;
		}
		
		visited[x][y] = false;
		return false;
		
	}

}

File: python/leetcode/WordSearch.py
class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def exist(self, board, word):
        m = len(board)
        if m == 0:
            return len(word) == 0
        n = len(board[0])
        if len(word) == 0:
            return True
        if len(word) > m * n:
            return False
        
        visited = [[False for a in range(n)] for b in range(m)]
        for j in range(n):
            for i in range(m):
                if board[i][j] == word[0] and self.dfs(i, j, board, word[1:], visited):
                    return True
        return False
        
    def dfs(self, x, y, board, word, visited):
        if word == '':
            return True
        m = len(board)
        n = len(board[0])
        visited[x][y] = True
        if x > 0 and board[x-1][y] == word[0] and not visited[x-1][y] and self.dfs(x-1, y, board, word[1:], visited):
            return True
        if x + 1 < m and board[x+1][y] == word[0] and not visited[x+1][y] and self.dfs(x+1, y, board, word[1:], visited):
            return True
        if y > 0 and board[x][y-1] == word[0] and not visited[x][y-1] and self.dfs(x, y-1, board, word[1:], visited):
            return True
        if y + 1 < n and board[x][y+1] == word[0] and not visited[x][y+1] and self.dfs(x, y+1, board, word[1:], visited):
            return True
        visited[x][y] = False
        return False
